compiler/syntax/src/res_core.ml

```diff
@@ -4885,11 +4885,49 @@ and parse_constr_decl_args p =
  *  | constr-name const-args
  *  | attrs constr-name const-args *)
 and parse_type_constructor_declaration_with_bar p =
+  let is_constructor_with_bar p =
+    Parser.lookahead p (fun state ->
+        match state.Parser.token with
+        | DocComment _ -> (
+          Parser.next state;
+          match state.token with
+          | Bar -> true
+          | _ -> false)
+        | Bar -> true
+        | _ -> false)
+  in
   match p.Parser.token with
-  | Bar ->
+  | _ when is_constructor_with_bar p ->
+    let doc_comment_attrs =
+      match p.Parser.token with
+      | DocComment (loc, s) ->
+        Parser.next p;
+        [doc_comment_to_attribute loc s]
+      | _ -> []
+    in
+    let has_doc_comment = not (doc_comment_attrs = []) in
     let start_pos = p.Parser.start_pos in
     Parser.next p;
-    Some (parse_type_constructor_declaration ~start_pos p)
+    let constr = parse_type_constructor_declaration ~start_pos p in
+    let pcd_loc =
+      if has_doc_comment then
+        {
+          constr.Parsetree.pcd_loc with
+          loc_end =
+            {
+              constr.pcd_loc.loc_end with
+              pos_lnum = constr.pcd_loc.loc_end.pos_lnum + 1;
+            };
+        }
```

Comment:
Without this line adjustment, something like

```rescript
type t =
  /** foo */
  | Foo
  /** bar */
  | Bar
  ```

will be printed as 
```rescript
type t =
  /** foo */
  | Foo

  /** bar */
  | Bar
  ```

because of this https://github.com/shulhi/rescript-compiler/blob/e71246b4315f9b3b8c3dc5acf7eb7fad6b95071e/compiler/syntax/src/res_printer.ml#L324-L328